packages/fury-adapter-oas3-parser/lib/parser/oas/parseOpenAPIObject.js

```diff
@@ -140,6 +143,14 @@ function parseOASObject(context, object) {
         }
       }
 
+      if (hosts) {
+        if (!hosts.isEmpty) {
+          api.push(new namespace.elements.Category(
+            hosts.content, { classes: ['hosts'] }
+          ));
+        }
```

Comment:
I'd suggest moving this logic into `parseServersArray` so that the `parseServersArray` will return a Category element if there are `servers`. That way here you can reduce the logic to only incorporate merging the hosts category into the `api` and we have something such as:

```js
if (hosts) {
   api.push(hosts);
}
```

This can be appended to the `pipeParseResult` part of `parseServers` (array).

This is for a couple of reasons:

- To make `parseOpenAPIObject` simpler so it can deal with as little logic as possible.
- Make it easier to test, because you can test this behaviour in the `parseServersArray` without testing the rest of the OpenAPI Object behaviours.
- To encourage re-usability, when another component needs to parse servers using `parseServersArray`, it already returns the structure it will need instead of additional processing.

Reply:
ok, changed

Reply:
@kylef 
I have a doubt about this because if in `parseServerArray` I change the function

```
function parseServersArray(context, element) {
  const { namespace } = context;

  const parseServers = pipeParseResult(namespace,
    parseArray(context, name, R.curry(parseServerObject)(context)));
  return parseServers(element);
}
```
to include this part, it will become something like
```
function parseServersArray(context, element) {
  const { namespace } = context;

  const parseServers = pipeParseResult(namespace,
    parseArray(context, name, R.curry(parseServerObject)(context)));

  return new namespace.elements.Category(
    parseServers(element), { classes: ['hosts'] }
  );
}
```
and, despite it works, the Categories are one level down, with `content` level that adds in between and a `content: parseResult` within the ServerArray. Could you please tell me how to change this function to not mess up the structure?

Reply:
@marcofriso you should be able to add another function argument to `pipeParseResult`, the output of `parseArray` on success will be passed to the next function before returning from `pipeParseResult`. This is similiar to what you've done in `parseServerObject`. I think something along the lines of:

```js
function parseServersArray(context, element) {
  const { namespace } = context;

  const parseServers = pipeParseResult(namespace,
    parseArray(context, name, R.curry(parseServerObject)(context)),
    (array) => {
      const category = new Category(array.content);
      category.classes.push('hosts');
      return category;
    });
  return parseServers(element);
}
```

Reply:
`pipeParseResult` behaves similiar to [Rambda pipe](https://ramdajs.com/docs/#pipe), in which it performs left-to-right function composition. The result of a left function is passed to the right function. In `pipeParseResult` that only happens if the parse result that `parseArray` returns contains a value so that it can "error-out".

Reply:
good,
in the end the function turns like this:

```js
function parseServersArray(context, element) {
  const { namespace } = context;

  const parseServers = pipeParseResult(namespace,
    parseArray(context, name, R.curry(parseServerObject)(context)),
    array => new namespace.elements.Category(
      array.content, { classes: ['hosts'] }
    ));

  return parseServers(element);
```